akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -218,3 +225,191 @@ private[fusing] object StreamOfStreams {
+    import PrefixAndTail._
+
+    private var left = if (n < 0) 0 else n
+    private var builder = Vector.newBuilder[T]
+    private var tailSource: TailSource[T] = null
+    private var tail: TailInterface[T] = null
+    builder.sizeHint(n)
+    private var pendingCompletion: MaterializationState = null
+
+    private val SubscriptionTimer = "SubstreamSubscriptionTimer"
+
+    private val onSubstreamPull = getAsyncCallback[Unit](_ ⇒ pull(in))
+    private val onSubstreamFinish = getAsyncCallback[Unit](_ ⇒ completeStage())
+    private val onSubstreamRegister = getAsyncCallback[TailInterface[T]] { tailIf ⇒
+      tail = tailIf
+      cancelTimer(SubscriptionTimer)
+      pendingCompletion match {
+        case NormalCompletion ⇒
+          tail.completeSubstream()
+          completeStage()
+        case FailureCompletion(ex) ⇒
+          tail.failSubstream(ex)
+          completeStage()
+        case _ ⇒
+      }
+    }
+
+    override protected def onTimer(timerKey: Any): Unit =
+      if (tailSource.materializationState.compareAndSet(NotMaterialized, TimedOut)) completeStage()
+
+    private def prefixComplete = builder eq null
+    private def waitingSubstreamRegistration = tail eq null
+
+    private def openSubstream(): Source[T, Unit] = {
+      val timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
+      tailSource = new TailSource[T](timeout, onSubstreamRegister.invoke, onSubstreamPull.invoke, onSubstreamFinish.invoke)
+      scheduleOnce(SubscriptionTimer, timeout)
+      builder = null
+      Source.fromGraph(tailSource)
+    }
+
+    // Needs to keep alive if upstream completes but substream has been not yet materialized
+    override def keepGoingAfterAllPortsClosed: Boolean = true
+
+    override def onPush(): Unit = {
+      if (prefixComplete) {
+        tail.pushSubstream(grab(in))
+      } else {
+        builder += grab(in)
+        left -= 1
+        if (left == 0) {
+          push(out, (builder.result(), openSubstream()))
+          complete(out)
+        } else pull(in)
+      }
+    }
+    override def onPull(): Unit = {
+      if (left == 0) {
+        push(out, (Nil, openSubstream()))
+        complete(out)
+      } else pull(in)
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (!prefixComplete) {
+        // This handles the unpulled out case as well
+        emit(out, (builder.result, Source.empty), () ⇒ completeStage())
+      } else {
+        if (waitingSubstreamRegistration) {
+          // Detach if possible.
+          // This allows this stage to complete without waiting for the substream to be materialized, since that
+          // is empty anyway. If it is already being registered (state was not NotMaterialized) then we will be
+          // able to signal completion normally soon.
+          if (tailSource.materializationState.compareAndSet(NotMaterialized, NormalCompletion)) completeStage()
```

Comment:
if the substream has been not materialized, we leave a tombstone in the child stream and we just go away. If it has been materialized, then we are fine, since registration comes soon and we can go away after the `onSubstreamRegister` callback.